src/System.Memory/src/System/SpanSortHelpers.Common.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+#if !netstandard
+using Internal.Runtime.CompilerServices;
+#endif
+
+namespace System
+{
+    internal static partial class SpanSortHelpersCommon
+    {
+        // This is the threshold where Introspective sort switches to Insertion sort.
+        // Empirically, 16 seems to speed up most cases without slowing down others, at least for integers.
+        // Large value types may benefit from a smaller number.
+        internal const int IntrosortSizeThreshold = 16;
+
+        internal static int FloorLog2PlusOne(int n)
+        {
+            Debug.Assert(n >= 2);
+            int result = 2;
+            n >>= 2;
+            while (n > 0)
+            {
+                ++result;
+                n >>= 1;
+            }
+            return result;
+        }
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Swap<T>(ref T items, int i, int j)
+        {
+            Debug.Assert(i != j);
+            Swap(ref Unsafe.Add(ref items, i), ref Unsafe.Add(ref items, j));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Swap<T>(ref T a, ref T b)
+        {
+            T temp = a;
+            a = b;
+            b = temp;
+        }
+
+        // This started out with just LessThan.
+        // However, due to bogus comparers, comparables etc.
+        // we need to preserve semantics completely to get same result.
+        internal interface IDirectComparer<in T>
+        {
+            bool GreaterThan(T x, T y);
+            bool LessThan(T x, T y);
+        }
+
+        //
+        // Type specific DirectComparer(s) to ensure optimal code-gen
```

Comment:
Are these type-specific structs necessary for optimal code-gen?

cc @AndyAyersMS

Reply:
@nietras can you compare perf with and without these extra comparers, and if it's quite different, show the key bits of generated code?

Also there still seems to be a healthy dose of `AggressiveInlining`. I realize it is time consuming to evaluate each of these on its own merits but we do need to be careful with framework code not to use this too broadly. I'm not asking you to remove them wholesale, but I would think they'd be less necessary at the outer edges of this and really only important in the innermost parts of sorting where we expect overhead costs to be magnified. So please think critically about where this attribute is needed. If I'm wrong and there is some benefit to having them higher up it would be good to see examples.

Also (not necessarily critiquing this change, just wondering aloud) it seems like the same algorithmic patterns are duplicated across different comparison contexts, and I wonder if there is some way to avoid this, or whether we are missing language features and/or concepts that would allow there to be one master version that gets appropriately specialized and optimized, or whether this is just how things are done for now. I know we haven't put the effort into optimizing general comparers the way we have for equality comparers -- maybe we should?

Reply:
@AndyAyersMS perf is significantly lower if using `IComparer` compared to the direct comparers. `Array.Sort` does the same for this, but via C++ template. The problem of course is `IComparer` will do more than one compare. I saw differences up to 2x for some scenarios. In fact, many scenarios could be improved if direct comparer was used for say `float` keys and `int` items/values. But due to the inconsistencies in sorting in `Array.Sort`, this would mean differences in sorting output.

So yes, these are necessary, and I would really like for perf to be good on existing runtimes.